### DISCUSSION TITLE
Future of the project

### DISCUSSION THREAD

**Discussion by cmsj** [Hammerspoon/hammerspoon#3506] [General]
Mostly for the attention of @asmagill and @latenitefilms , but general discussions are very welcome.

I think it's been pretty clear for a while now that I am not doing a good job at maintaining Hammerspoon. If I'm completely honest with myself, I just don't find any joy anymore in working on this codebase.

I'm super proud of what we've all done together over the last 8.5 years, but I think we should talk about what happens next.

I'm still very hungry for programming, but I find myself wanting to play with new things and refine architectures until I think they're excellent. Hammerspoon has enough of a userbase, and enough very old and crufty code, that it feels impossible to bring the kinds of changes I think it needs (how many times have we discussed what we'd like to do before the 1.0 release that seems just as far away as it did a few years ago).

I'm also still very much interested in automation - I still use Hammerspoon all the time and I don't want to be without it, which means I can't just abandon it and let it slide until it no longer works properly.

If I had infinite free time, I would start from scratch and focus on:

- Swift instead of Objective C
- As little Lua in the main codebase as possible (or possibly even switch to something else - I've never been a huge fan of Lua)
- Watchables
- Figure out if it's possible to let users make UIs (I recently gave up making a complex config for Stream Deck + OBS + MIDI + lighting, because it was 20x quicker to just use Elgato's Stream Deck software)
- Much more API consistency

Unfortunately, I don't have infinite free time, so on my own such an undertaking would take an extremely long time.

All of which brings us to the part where I need other opinions and voices - is anyone else feeling like this? Is anyone else hungry for a change? Would anyone else be able to spare time to work on such a thing with me?

We would be signing up for a ton of work and a 100% guarantee that we would require every single existing Hammerspoon Spoon and config to be at least somewhat rewritten. I would feel bad about that, but I'd rather feel bad about it and still be actively working on the project, than not.

**Comment by latenitefilms** [comment]
I'm still very keen to continue to maintain and keep Hammerspoon alive, as it's the "engine" behind [CommandPost](https://commandpost.io), and I have no intention of killing CommandPost anytime soon.

[CommandPost](https://commandpost.io) has grown a lot over the years if you haven't checked it out recently. It's pretty amazing what we've been able to build using Lua and simple HTML user interfaces!

Having said, that, I've spent more of this year building commercial/paid Mac apps:

- **[BRAW Toolbox](https://brawtoolbox.io/)** allows you to import Blackmagic RAW files directly into Final Cut Pro without transcoding. Written in Objective-C. Currently on Mac App Store.
- **[Gyroflow Toolbox](https://gyroflowtoolbox.io/)** allows you to import Gyroflow Projects into Apple's Final Cut Pro. Written in Objective-C. Currently on Mac App Store.
- **[Transfer Toolbox](https://transfertoolbox.io/)** allows you to convert Final Cut Pro (for Mac) libraries into Final Cut Pro (for iPad) projects. Written in Swift + Swift UI. Currently stuck in TestFlight as Apple REALLY don't want to approve it.
- **[Recall Toolbox](https://recalltoolbox.io/)** is a Final Cut Pro Workflow Extension that acts as a Shared Pasteboard. Written in Swift + Swift UI. Currently in development on TestFlight. Hoping to release soon.
- **[Marker Toolbox](https://markertoolbox.io/)** allows you to import Vimeo CSV, Dropbox Replay JSON & Email Comments into Apple's Final Cut Pro.

So, I now know enough Objective-C and Swift + Swift UI to be very dangerous, haha.

It's worth noting that NONE of the above would have been remotely possible without the help and support of @cmsj & @asmagill - I owe all my Objective-C knowledge and skills to you two - you've both helped me so incredibly much. Thank you, thank you.

Throwing it out there... I would be totally open to "taking over" the Hammerspoon repo and maintaining it. I'd probably rebuild the website to be more like [CommandPost](https://commandpost.io) and replace the icon to [match our existing icons](https://latenitefilms.com/technology/) (ignoring the last 3 as they're just placeholders). Apart from that, I'd just try and keep it alive.

I'm also TOTALLY excited and interested in helping you build **Hammerspoon Pro** at the same time - a Swift + Swift UI re-write of Hammerspoon. I would LOVE to help make this happen. You could even drop Lua all together, and make the scripting in Swift?

**Reply by latenitefilms** [reply]
Another option would be to "port" the Hammerspoon Lua engine over to something like [Raycast](https://www.raycast.com) as use that as the Interface?

**Reply by cmsj** [reply]
I'm not sure that scripting in Swift is possible, although technically you could compile the user's Swift config and load it as a library, you'd just have to restart the app on every config change.

I hadn't spotted that you've been working on other apps as well - that's awesome!

If we were to do the rewrite, would you consider porting CommandPost to it?

**Reply by latenitefilms** [reply]
I was more thinking we should use "Swift" as our scripting language, but process it in the same way as Lua kinda works. Obviously a lot of work - but could be an interesting take? Or just keep it simple and stick with Lua, as I do love it, haha.

If we made **Hammerspoon Pro**, I'd probably consider making **CommandPost Pro** alongside it, and hope it gets to the point where we can abandon the older versions eventually. Maybe even make it a paid app, as the whole community funded sponsorship model never really worked.

**Reply by latenitefilms** [reply]
It's also worth noting that most of my "Toolbox" apps are taking features out of CommandPost and making them standalone paid apps.

**Reply by cmsj** [reply]
I'm not sure I want to get into the business of making my own scripting language that happens to look like Swift, that sounds like a ton of extra work!

I hadn't really thought about whether to keep it open source or not. I've always assumed there isn't enough of a market for a very nerdy automation tool like Hammerspoon, to charge for it.

**Reply by latenitefilms** [reply]
Based on how much money I've made from all my apps, I definitely don't think it's a big money making venture, but even a small amount of income could be helpful to devote more time to it. For me personally, the more money I make off software, the more time I can justify coding - as it's not my "real job". I'd be more than happy to collaborate with you on a closed-source model, or an open-source model.

**Reply by cmsj** [reply]
I think I would prefer open source, but something I would want to change from 1.0 is to not have all the dylibs - rather, I'd want to put our functionality in a single HammerKit.framework, which would then make it really easy to make little task-specific closer source apps like yours, by just using that framework.

I have a fun dream that eventually HammerKit would be so useful and so correct, that *other people* would want to do the same, and would give us features/fixes back, which almost makes me feel like HammerKit should be LGPL.

**Reply by latenitefilms** [reply]
Totally agree - and LOVE idea of a Swift-based HammerKit Framework. All the magic should happen in there, with minimal stuff happening in Lua.

I'm totally happy to trust you in terms of licensing, but I don't have a problem with keeping it MIT.

**Reply by cmsj** [reply]
Cool cool :)

**Comment by latenitefilms** [comment]
Just thinking out loud if you did want to handball Hammerspoon to me as the main maintainer, these are the initial steps I'd probably do (with your blessing/advice of course):

- Get @cmsj to transfer ownership of [hammerspoon.org](https://hammerspoon.org) to [LateNite Films](https://latenitefilms.com)
- I'd redirect [hammerspoon.org](https://hammerspoon.org) to [hammerspoon.io](https://hammerspoon.io), and make that the default domain (to keep in line with my other apps)
- Change the bundle ID from `org.hammerspoon.Hammerspoon` to `com.latenitefilms.Hammerspoon`
- I'd get [Matthew Skiles](https://matthewskiles.com) to create a more modern icon in same design
- I'd redo the website using [Retype](https://retype.com) like [commandpost.io](https://commandpost.io)
- Close all the existing GitHub issues, explaining we're pushing for v1.0.0 and to re-open if still relevant
- I'd push out v1.0.0 with these changes
- I'd push out more regular updates when external Pods are updated

Apart from the above, I don't think anything else would really change initially. I'd still just rely on the community to add pull requests for new extensions and bug fixes.

Thoughts?

**Reply by cmsj** [reply]
@latenitefilms that's an interesting model - kinda put 1.0 immediately into maintenance mode. If we go a route like that, we arguably wouldn't even need to transfer the ownership, and we could just get cracking on 2.0.

**Reply by latenitefilms** [reply]
Obviously I'm TOTALLY happy and would encourage you to continue to be the main guru behind Hammerspoon - I feel like you're a much more sensible and responsible grown up than me!

I feel like Hammerspoon is kinda already in maintenance mode, no one's really added that many big new feature recently, and the last few updates have been basically minor bug fixes.

I'm totally happy to get started on Hammerspoon Pro/v1/v2. Maybe the best way forward is to create a new Hammerspoon Pro repo and start documenting goals, what we want to achieve, minimal feature set, etc.?

**Reply by cmsj** [reply]
Yep, I like the sound of that. Thanks!

**Comment by asmagill** [comment]
Not sure how I missed this the first time around... I understand where you're coming from @cmsj -- while I have a few specific things I'd still like to accomplish (Catalina back-port, revamp webview and canvas so that it's easier to add other NSView based things without replicating a crap-ton of code, proper regex, etc.) I do agree that the motivation and drive isn't what it once was.

And I'd love to get Lua/LuaSkin (or whatever the scripting engine becomes) *off* of the main thread and let the main thread just be a queue for tasks that are supposed to occur there (UI updates, eventtaps, etc.) while other threads do the heavy lifting. But that's basically a full rewrite...

I'm not adverse to @latenitefilms taking over as head. I still hope to contribute, but it will be more of a "as and when I can" and not necessarily on a regular schedule.

**Reply by latenitefilms** [reply]
This is pretty cool:

https://community.folivora.ai/t/floating-menu-alpha-development-thread/31845

**Reply by latenitefilms** [reply]
Lots of interesting stuff in here too:

https://developer.apple.com/documentation/macos-release-notes/appkit-release-notes-for-macos-14

**Comment by cmsj** [comment]
Another change I'm making is adding a Discord server because IRC has proven to be almost entirely useless. Invite link is: https://discord.gg/vxchqkRbkR

**Comment by asmagill** [comment]
@cmsj, I know Hammerspoon itself has some dependancies that rely specifically on you (cert signing, Sparkle feed, etc.) but for the spoons, is this something I could also do? I think we have a few that are probably ready for inclusion, but I don't know the steps you follow to create the zip file, verify the docs, etc... is it a script or does it also rely on tokens or certificates only you have?

Not that I promise specific regularity, but it might lessen the load that depends on you at the moment.

**Reply by cmsj** [reply]
Assuming everything is in good shape, the process is fully automated, so yes, you could merge a Spoon PR and everything should update properly with the zips and website docs. The "Actions" tab lets you check the outcome.

I would definitely welcome you helping out with the Spoons repo!

**Reply by asmagill** [reply]
Traveling at the moment and have less access than I thought I would. I’ll give it a try this weekend and let you know if I run into any issues!

**Reply by roeeyn** [reply]
Is there any way I can help? I've been using Lua for a while, and I'd love to help in this new stage of Hammerspoon.

**Reply by asmagill** [reply]
@cmsj, just noticed you merged a spoon and apologize that my stepping up to do so has been delayed... unexpected family concerns, but I am hoping to this week.

 I just want to confirm (and after looking at the code in the actions, I am pretty sure...) that all I need to do is make sure the spoon src is in the right directory and then merging it will take care of the docs and zip file, correct? Should included docs or zip files be removed manually first, or will they be overwritten?

**Reply by cmsj** [reply]
@asmagill its a good idea to make sure the test action has run on the PR first, but after that, yep, just merge. The docs and zip will be recreated by the merge action.

If any of that automation fails, it's a bug that I should fix 😁

**Reply by cmsj** [reply]
> Is there any way I can help? I've been using Lua for a while, and I'd love to help in this new stage of Hammerspoon.

@roeeyn hard to say at this stage. I've mostly just been thinking about how a Swift version of LuaSkin can benefit from Swift language features to simplify things, I've not actually written very much code yet.

**Comment by 18o** [comment]
any plan to provide 3rd language api? such as Rust
is it possible integrate python3 instead of Lua ? in this way there will be a lot of 3rd libs can be used.

**Comment by muescha** [comment]
I never liked Lua. So an scripting engine for user scripts in Swift or TypeScript would attract more people, because they are more popular than Lua.

**Comment by ghost** [comment]
I would be happy with scripting in Python or JavaScript — either Deno or Node.js. Deno is probably easier to integrate. I've had to completely rewrite many of my Python CLI scripts in Lua to integrate them well with Hammerspoon... and yes, it was a pain because Lua is a very limited language for scripting.
